File: download_assets.py
import os
import json
import urllib.request
import urllib.parse
import sys

BASE_URL = "https://xn--2v5bo7x.com"

# 폴더 경로 정의
PATHS = [
    "static/css",
    "static/js",
    "static/images",
    "static/images/EquipImage",
    "data/events"
]

def create_directories():
    print("[-] 디렉토리를 생성하는 중...")
    for path in PATHS:
        os.makedirs(path, exist_ok=True)
    print("[+] 디렉토리 생성 완료.")

def download_file(url, filepath):
    try:
        # 공백이나 특수 문자가 포함된 URL 인코딩 처리
        parsed_url = urllib.parse.urlparse(url)
        encoded_path = urllib.parse.quote(parsed_url.path)
        encoded_url = urllib.parse.urlunparse(parsed_url._replace(path=encoded_path))
        
        req = urllib.request.Request(
            encoded_url, 
            headers={'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64)'}
        )
        with urllib.request.urlopen(req) as response, open(filepath, 'wb') as out_file:
            out_file.write(response.read())
        print(f"[+] 다운로드 성공: {filepath}")
        return True
    except Exception as e:
        print(f"[!] 다운로드 실패: {url} -> {e}")
        return False

def replace_colon_on_item_name(name):
    return name.replace(' : ', '-').replace(': ', '-').replace(':', '-')

def main():
    create_directories()
    
    # 1. CSS 파일 다운로드
    css_files = [
        "style.css",
        "common-modern.css",
        "darkmode.css",
        "randbox.css",
        "modern-index.css",
        "alerts.css"
    ]
    print("\n[-] CSS 파일 다운로드 중...")
    for css in css_files:
        download_file(f"{BASE_URL}/static/css/{css}", f"static/css/{css}")
        
    # 2. JS 파일 다운로드
    js_files = [
        "darkmode.js",
        "common-modern.js",
        "auth.js",
        "randbox.js"
    ]
    print("\n[-] JS 파일 다운로드 중...")
    for js in js_files:
        download_file(f"{BASE_URL}/static/js/{js}", f"static/js/{js}")
        
    # 3. 기타 에셋 다운로드
    print("\n[-] 기본 에셋 다운로드 중...")
    download_file(f"{BASE_URL}/static/images/favicon.ico", "static/images/favicon.ico")
    download_file(f"{BASE_URL}/static/images/noimg.webp", "static/images/noimg.webp")
    
    # 4. 이벤트 목록 다운로드
    print("\n[-] 이벤트 목록 다운로드 중...")
    events_list_path = "data/events.json"
    if download_file(f"{BASE_URL}/list_event_files", events_list_path):
        with open(events_list_path, 'r', encoding='utf-8') as f:
            events = json.load(f)
            
        unique_items = set()
        
        # 5. 각 이벤트의 개별 상세 데이터 다운로드
        print("\n[-] 개별 이벤트 상세 데이터 및 아이템 이미지 수집 중...")
        for event in events:
            event_id = event['id']
            # URL에 특수문자 및 공백 처리
            encoded_id = urllib.parse.quote(event_id)
            event_url = f"{BASE_URL}/get_event_data/{encoded_id}"
            event_file_path = f"data/events/{event_id}.json"
            
            if download_file(event_url, event_file_path):
                # 아이템 이름 수집
                try:
                    with open(event_file_path, 'r', encoding='utf-8') as ef:
                        event_data = json.load(ef)
                        for box in event_data.get('boxes', []):
                            for item in box.get('items', []):
                                if item.get('name'):
                                    unique_items.add(item['name'])
                except Exception as ex:
                    print(f"[!] {event_id} 데이터 파싱 오류: {ex}")
        
        # 6. 수집된 모든 고유 아이템 이미지 다운로드
        print(f"\n[-] 수집된 고유 아이템 개수: {len(unique_items)}")
        print("[-] 아이템 이미지 다운로드 중...")
        
        success_count = 0
        for item_name in unique_items:
            img_filename = replace_colon_on_item_name(item_name)
            img_url = f"{BASE_URL}/static/images/EquipImage/{img_filename}.webp"
            img_path = f"static/images/EquipImage/{img_filename}.webp"
            
            # 파일이 이미 존재하면 다운로드 건너뛰기 (시간 절약)
            if os.path.exists(img_path):
                success_count += 1
                continue
                
            if download_file(img_url, img_path):
                success_count += 1
                
        print(f"\n[+] 자산 다운로드 완료! 총 {len(unique_items)}개 중 {success_count}개 다운로드 성공.")
    else:
        print("[!] 이벤트 목록을 받아오지 못해 이미지 다운로드를 진행할 수 없습니다.")

if __name__ == "__main__":
    main()

File: download_assets.js
const fs = require('fs');
const path = require('path');
const https = require('https');

const BASE_URL = "https://xn--2v5bo7x.com";

// 폴더 구조 생성
const PATHS = [
    "static/css",
    "static/js",
    "static/images",
    "static/images/EquipImage",
    "data/events"
];

function createDirectories() {
    console.log("[-] 디렉토리를 생성하는 중...");
    for (const p of PATHS) {
        fs.mkdirSync(p, { recursive: true });
    }
    console.log("[+] 디렉토리 생성 완료.");
}

function downloadFile(url, filepath) {
    return new Promise((resolve) => {
        try {
            const finalUrl = new URL(url);

            https.get(finalUrl, {
                headers: { 'User-Agent': 'Mozilla/5.0' }
            }, (res) => {
                if (res.statusCode !== 200) {
                    console.log(`[!] 다운로드 실패 (${res.statusCode}): ${url}`);
                    resolve(false);
                    return;
                }

                const fileStream = fs.createWriteStream(filepath);
                res.pipe(fileStream);

                fileStream.on('finish', () => {
                    fileStream.close();
                    console.log(`[+] 다운로드 성공: ${filepath}`);
                    resolve(true);
                });

                fileStream.on('error', (err) => {
                    console.log(`[!] 파일 쓰기 실패: ${filepath} -> ${err.message}`);
                    resolve(false);
                });
            }).on('error', (err) => {
                console.log(`[!] 네트워크 연결 실패: ${url} -> ${err.message}`);
                resolve(false);
            });
        } catch (e) {
            console.log(`[!] URL 파싱 실패: ${url} -> ${e.message}`);
            resolve(false);
        }
    });
}

function replaceColonOnItemName(name) {
    return name.replace(/ : /g, '-').replace(/: /g, '-').replace(/:/g, '-');
}

async function main() {
    createDirectories();

    // 1. CSS 파일 다운로드
    const cssFiles = [
        "style.css",
        "common-modern.css",
        "darkmode.css",
        "randbox.css",
        "modern-index.css",
        "alerts.css"
    ];
    console.log("\n[-] CSS 파일 다운로드 중...");
    for (const css of cssFiles) {
        await downloadFile(`${BASE_URL}/static/css/${css}`, `static/css/${css}`);
    }

    // 2. JS 파일 다운로드
    const jsFiles = [
        "darkmode.js",
        "common-modern.js",
        "auth.js",
        "randbox.js"
    ];
    console.log("\n[-] JS 파일 다운로드 중...");
    for (const js of jsFiles) {
        await downloadFile(`${BASE_URL}/static/js/${js}`, `static/js/${js}`);
    }

    // 3. 기타 공통 에셋 다운로드
    console.log("\n[-] 기본 에셋 다운로드 중...");
    await downloadFile(`${BASE_URL}/static/images/favicon.ico`, "static/images/favicon.ico");
    await downloadFile(`${BASE_URL}/static/images/noimg.webp`, "static/images/noimg.webp");

    // 4. 이벤트 목록 다운로드
    console.log("\n[-] 이벤트 목록 다운로드 중...");
    const listSuccess = await downloadFile(`${BASE_URL}/list_event_files`, "data/events.json");

    if (listSuccess) {
        const events = JSON.parse(fs.readFileSync("data/events.json", "utf8"));
        const uniqueItems = new Set();

        console.log("\n[-] 개별 이벤트 상세 데이터 수집 중...");
        for (const ev of events) {
            const eventId = ev.id;
            const eventUrl = `${BASE_URL}/get_event_data/${eventId}`;
            const eventPath = `data/events/${eventId}.json`;

            const success = await downloadFile(eventUrl, eventPath);
            if (success) {
                try {
                    const eventData = JSON.parse(fs.readFileSync(eventPath, "utf8"));
                    for (const box of eventData.boxes || []) {
                        for (const item of box.items || []) {
                            if (item.name) {
                                uniqueItems.add(item.name);
                            }
                        }
                    }
                } catch (e) {
                    console.log(`[!] ${eventId} JSON 파싱 실패: ${e.message}`);
                }
            }
        }

        console.log(`\n[+] 수집 완료. 고유 아이템 개수: ${uniqueItems.size}`);
        console.log("[-] 아이템 이미지 다운로드 시작...");

        let successCount = 0;
        const itemArray = Array.from(uniqueItems);
        
        // 너무 많은 요청을 동시에 보내서 에러가 나지 않도록 순차 처리하되, 속도를 위해 청크 단위로 병렬 수행
        const CHUNK_SIZE = 10;
        for (let i = 0; i < itemArray.length; i += CHUNK_SIZE) {
            const chunk = itemArray.slice(i, i + CHUNK_SIZE);
            const promises = chunk.map(async (name) => {
                const imgFilename = replaceColonOnItemName(name);
                const imgUrl = `${BASE_URL}/static/images/EquipImage/${imgFilename}.webp`;
                const imgPath = `static/images/EquipImage/${imgFilename}.webp`;

                if (fs.existsSync(imgPath)) {
                    successCount++;
                    return;
                }

                const downloaded = await downloadFile(imgUrl, imgPath);
                if (downloaded) {
                    successCount++;
                }
            });
            await Promise.all(promises);
        }

        console.log(`\n[+] 모든 리소스 다운로드 완료! 총 ${uniqueItems.size}개 중 ${successCount}개 이미지 확보 완료.`);
    } else {
        console.log("[!] 이벤트 목록 수집 실패.");
    }
}

main();
